# LongPhase-S
LongPhase-S is a program for phasing and haplotagging germline and somatic variants from tumor-normal pair long-read sequencing data, based on [**LongPhase**](https://github.com/twolinin/longphase).

---
- [Installation](#installation)
- [Usage](#usage)
	- [Phase command](#phase-command)
		- [SNP-only phasing](./docs/phase.md#snp-only-phasing)
		- [SNP and indel co-phasing](./docs/phase.md#snp-and-indel-co-phasing)
		- [SNP and SV co-phasing](./docs/phase.md#snp-and-sv-co-phasing)
		- [SNP and modification co-phasing](./docs/phase.md#snp-and-modification-co-phasing)
		- [The complete list of phase parameters](./docs/phase.md#the-complete-list-of-phase-parameters)
		- [Output of SNP and indel phasing](./docs/phase.md#output-of-snp-and-indel-phasing)
		- [Output files of SNP and SV co-phasing](./docs/phase.md#output-files-of-snp-and-sv-co-phasing)
	- [Haplotag command](#haplotag-command)
		- [The complete list of haplotag parameters](./docs/haplotag.md#the-complete-list-of-haplotag-parameters)
	- [Somatic haplotag command](#somatic-haplotag-command)
		- [SNP-only somatic haplotag](./docs/somatic_haplotag.md#snp-only-somatic-haplotag)
		- [Somatic haplotag benchmark](./docs/somatic_haplotag.md#somatic-haplotag-benchmark)
		- [The complete list of somatic haplotag parameters](./docs/somatic_haplotag.md#the-complete-list-of-somatic-haplotag-parameters)
		- [Ouputfiles](#output-files)
			- [Tumor purity prediction file](./docs/somatic_haplotag.md#tumor-purity-prediction-result-file)
			- [Somatic calling result VCF file](./docs/somatic_haplotag.md#somatic-calling-result-vcf-file)
			- [Benchmark metrics file](./docs/somatic_haplotag.md#benchmark-metrics-file)
	- [Predict purity command](#predict_purity-command)
		- [Complete list of purity prediction parameters](./docs/predict_purity.md#complete-list-of-purity-prediction-parameters)
		- [Tumor purity prediction file](./docs/somatic_haplotag.md#tumor-purity-prediction-result-file)

- [Input Preparation](#input-preparation)
	- [Generate reference index](./docs/input_preparation.md#generate-reference-index)
	- [Generate alignment and index files](./docs/input_preparation.md#generate-alignment-and-index-files)
	- [Generate single nucleotide polymorphism (SNP) file](./docs/input_preparation.md#generate-single-nucleotide-polymorphism-snp-file)
	- [Generate Structural variation (SV) file](./docs/input_preparation.md#generate-structural-variation-sv-file)
 	- [Carry methylation tags to BAMs](./docs/input_preparation.md#carry-methylation-tags-to-bams)
- [Comparison with other SNP-phasing programs](#comparison-with-other-snp-phasing-programs)
- [Citation](#citation)
- [Contact](#contact)
---
## Installation
You are recommended to download a [linux 64bit binary release](https://github.com/twolinin/longphase/releases/download/v1.7.3/longphase_linux-x64.tar.xz) without compilation. 

```
wget https://github.com/twolinin/longphase/releases/download/v1.7.3/longphase_linux-x64.tar.xz
tar -xJf longphase_linux-x64.tar.xz
```

An executable file, longphase_linux-x64, can be executed directly. If you need to compile a local version, you can clone and compile using the following commands, and make sure that the environment has zlib installed. If you require setting up a virtual environment, we also provide a [Dockerfile](https://github.com/twolinin/longphase/blob/main/Dockerfile).

```
git clone https://github.com/twolinin/longphase.git
cd longphase
autoreconf -i
./configure
make -j 4
```

---
## Usage
### Phase command
#### SNP-only phasing
For SNP-only phasing, the input of LongPhase consists of SNPs in VCF (e.g., SNP.vcf), an indexed reference in Fasta (e.g., reference.fasta, reference.fasta.fai), and one (or multiple) indexed read-to-reference alignment in BAM (e.g., alignment1.bam, alignment1.bai, alignment2.bam, ...) (see [Input Preparation](#input-preparation)). The users should specify the sequencing platform (--ont for Nanopore and --pb for PacBio). An example of SNP phasing usage is shown below.
```
longphase-s phase \
-s SNP.vcf \
-b alignment1.bam \
-b alignment2.bam \
-r reference.fasta \
-t 8 \
-o phased_prefix \
--ont # or --pb for PacBio Hifi
```
- [The complete list of phase parameters](./docs/phase.md#the-complete-list-of-phase-parameters)
- [Output of SNP and indel phasing](./docs/phase.md#output-of-snp-and-indel-phasing)
- [Output files of SNP and SV co-phasing](./docs/phase.md#output-files-of-snp-and-sv-co-phasing)

### Haplotag command
This command tags (assigns) each read (in BAM) to one haplotype in the phased SNP/SV VCF. i.e., reads will be tagged as HP:i:1 or HP:i:2. In addition, the haplotype block of each read is stored in the PS tag. The phased VCF can be also generated by other programs as long as the PS or HP tags are encoded. The author can specify ```--log``` for additionally output a plain-text file containing haplotype tags of each read without parsing BAM.
```
longphase-s haplotag \
-r reference.fasta \
-s phased_snp.vcf \
--sv-file phased_sv.vcf \
-b alignment.bam \
-t 8 \
-o tagged_bam_prefix
```
- [The complete list of haplotag parameters](./docs/haplotag.md#the-complete-list-of-haplotag-parameters)

### Somatic haplotag command
#### SNP-only somatic haplotag
This command performs tumor purity prediction and somatic variant calling using tumor-normal pair BAM files, then tags (assigns) each read in tumor BAM to one haplotype based on phased normal SNP VCF and tumor SNP VCF. The reads will be tagged as:
- HP:z:1 or HP:z:2 for reads with germline SNPs
- HP:z:H1-1 or HP:z:H2-1 for reads with somatic SNPs derived from germline haplotype 1 or 2
- HP:z:H3 for reads with somatic SNPs that cannot be derived from germline haplotypes

In addition, the haplotype block of each read is stored in the PS tag (only for reads with phased SNPs). The phased VCF can be generated by other programs as long as the PS or HP tags are encoded. The author can specify `--log` for additionally output a plain-text file containing haplotype tags of each read without parsing BAM.

```
longphase-s somatic_haplotag \
-s phased_normal_snp_vcf \
-b normal_bam \
--tumor-snp-file tumor_snp_vcf \
--tumor-bam-file tumor_bam \
-r reference.fasta \
-t thread_numbers \
-o tagged_tumor_bam_prefix \
--tagSupplementary \
-q 20 \
```
- [The complete list of somatic haplotag parameters](./docs/somatic_haplotag.md#the-complete-list-of-somatic-haplotag-parameters)

#### Output files:
- Tagged tumor BAM file
- [Tumor purity prediction file](./docs/somatic_haplotag.md#tumor-purity-prediction-file)
- [Somatic calling result VCF](./docs/somatic_haplotag.md#somatic-calling-result-vcf-file) (if `--output-somatic-vcf` is enabled)
- [Benchmark metrics file](./docs/somatic_haplotag.md#benchmark-metrics-file) (if truth files are provided)

#### Somatic haplotagging benchmark :
- If `--truth-vcf` is provided, it will evaluate the performance of somatic haplotagging by comparing the reads that truly contain somatic variants with the reads that are tagged as somatic reads. 
- If `--truth-bed` is also provided, the evaluation will only consider variants within these regions.
- [See detailed benchmark methodology](./docs/somatic_haplotag.md#somatic-haplotagging-benchmark)

### Predict_purity command
This command predicts tumor purity using tumor-normal pair BAM and VCF files, along with haplotype information, and outputs the prediction file.
```
longphase-s predict_purity \
-s phased_normal_vcf \
-b normal_bam \
--tumor-snp-file tumor_vcf \
--tumor-bam-file tumor_bam \
-r reference_fasta \
-t thread_numbers \
-o output_predix
```
- [Complete list of purity prediction parameters](./docs/predict_purity.md#complete-list-of-purity-prediction-parameters)
- [Tumor purity prediction file](./docs/somatic_haplotag.md#tumor-purity-prediction-file)
---
## Input Preparation
- [Generate reference index](./docs/input_preparation.md#generate-reference-index)
- [Generate alignment and index files](./docs/input_preparation.md#generate-alignment-and-index-files)

---
## Comparison with other SNP-phasing programs
LongPhase is >30x faster than WhatsHap and Margin and produces much larger blocks when tested on HG002, HG003,and HG004.
![btac058f3](https://github.com/twolinin/longphase/assets/6086073/af3a75a1-6268-4700-9dcc-4a6f34e86f7a)

## Speed
LongPhase can phase a human genome within 1-2 minutes.
phase (-t 24) | v1.6 (Time) | v1.6 (Memory)
-- | -- | -- 
HG002 ONT R10.4.1 10x |  39s | 15.1G
HG002 ONT R10.4.1 20x |  53s | 15.6G
HG002 ONT R10.4.1 30x |  68s | 24.4G
HG002 ONT R10.4.1 40x |  217s | 26.6G
HG002 ONT R10.4.1 50x |  262s | 22.2G
HG002 ONT R10.4.1 60x |  113s | 33.4G


---
## Citation
Jyun-Hong Lin, Liang-Chi Chen, Shu-Qi Yu and Yao-Ting Huang, [LongPhase: an ultra-fast chromosome-scale phasing algorithm for small and large variants](https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btac058/6519151), Bioinformatics, 2022.

---
## Contact
Yao-Ting Huang, ythuang at cs.ccu.edu.tw
